In [504]:
import pandas as pd

In [ ]:
## !pip install jinja2
##!pip install the matplotlib

In [505]:
from datetime import datetime
import numpy as np

# Ejercicio 1


Cargar el fichero excel

In [497]:
df = pd.read_excel('E:/osley/Barecelona activa/curso especializacion analisis de datos/sprint 8/sprint10_complex.xlsx', header=3)

In [439]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1007 entries, 0 to 1006
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Unnamed: 0                  1007 non-null   int64  
 1   Nom                         1006 non-null   str    
 2   Cognoms                     1006 non-null   str    
 3   DNI                         1006 non-null   str    
 4   País d'origen               1006 non-null   str    
 5   Ciutat                      1006 non-null   str    
 6   Dia de Naixement            1006 non-null   float64
 7   Mes de Naixement            1006 non-null   float64
 8   Any de Naixement            1006 non-null   float64
 9   Gènere                      966 non-null    str    
 10  Salari mensual              1006 non-null   str    
 11  Fills                       419 non-null    object 
 12  No Fills                    585 non-null    object 
 13  Grup Professional           1005 non-null   

El dataset tiene 1007 registros y 20 columnas, tipos de datos float64(7), int64(1), object(3), str(9)

In [440]:
display(df.head(10))

,Unnamed: 0,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,CÃ rrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
0,0,Joana,Gil Navarro,66722344X,Espanya,Valladolid,23.0,3.0,1958.0,Dona,"1,469 €",True,NaN,Grup B,Cap de projecte,3.0,True,32108.0,25.0,10.1
1,1,Marc,Muñoz,48840994W,Espanya,Alacant,8.0,11.0,1960.0,H,"2,718 €",NaN,True,Grup C,Senior analyst,0.0,True,19496.0,10.4,18.7
2,2,Noa,Serra,14308421X,Espanya,Alacant,27.0,4.0,1961.0,D,1358 euros,True,NaN,Grup A,Tècnic IT,4.0,NaN,NaN,NaN,16.7
3,3,Pol,Gil,58586340F,Espanya,Sevilla,13.0,10.0,1985.0,H,"1,478 €",True,NaN,Grup B,Data Analyst,2.0,False,NaN,NaN,18.3
4,4,David,Vila,82070937P,Espanya,Bilbao,29.0,11.0,1965.0,Dona,"1,284 €",NaN,True,Grup B,Administratiu,NaN,True,NaN,11.8,13.1
5,5,Alexia,Puig,25014631B,Espanya,Las Palmas de Gran Canaria,1.0,8.0,1978.0,D,"2,069 €",NaN,True,Grup B,Cap de projecte,0.0,NaN,NaN,NaN,19.9
6,6,David,Gómez Vila,36998038E,Espanya,Palma,25.0,9.0,1971.0,D,"2,349 €",NaN,True,Grup B,Cap de projecte,NaN,True,6196.0,10.6,17.8
7,7,Carmen,López,47376585K,Espanya,Barcelona,8.0,9.0,1962.0,H,"2,855 €",True,NaN,Grup C,Responsable de vendes,4.0,True,20000.0,5.9,15.4
8,540,Nora,Andersen Pedersen,37289414J,Noruega,Stavanger,8.0,3.0,1963.0,D,"1,338 €",NaN,True,Grup B,Analista junior,0.0,False,NaN,NaN,7.6
9,9,Alba,García,75181648S,Espanya,Las Palmas de Gran Canaria,32.0,0.0,1940.0,H,"1,450 €",True,NaN,Grup A,Senior analyst,4.0,False,NaN,NaN,20.1


Hay filas vacias, filas con datos inconguentes, errores tipograficos, celdas sin datos, generos invertidos, fechas incorrectas, registros duplicados, tipos de datos que hay que cambiar
Hay una Columna que tine la misma numeracion que el indice, varias colunnas que se pueden representar en un una sola columna(columnas de tiempo y de hijos). 

# Ejercicio 2

Eliminacion de columna 'Unnamed' que tiene los mis valores que el indice

In [506]:
if 'Unnamed: 0' in df.columns:
    df = df.drop('Unnamed: 0', axis=1)

Eliminar la fila 1006 por tener todos sus valores Nan

In [507]:
df.dropna(how='all', inplace=True)
df = df.drop(1006)

Arreglar el nombre de la columna CÃ\xa0rrec' por 'Carrec'

In [508]:
df.rename(columns={'CÃ\xa0rrec': 'Carrec'}, inplace=True)
df.dropna(how='all', inplace=True)

Definimos dicionario con los errores tipograficos y su corección

In [509]:
mapeo_caracteres = {
    'Ã\xa0': 'à', 'é': 'é', 'Ã\xad': 'í', 'ú': 'ú', 'É': 'Á', 
    'ñ': 'ñ', 'ó': 'ó', 'á': 'á', 'è': 'e', 'Ã\x81': 'A', 
    'ä': 'ä', 'ü': 'ü', 'ö': 'ö', 'ç': 'ci', 'ø': 'ø'
}


Función de limpieza de texto optimizada con diccionario

Asegurar tipo string y reemplazar todo el diccionario en un solo paso

In [510]:


def corregir_subcadenas_campo(dataframe, columnas_a_corregir, diccionario_cambios):
    for col in columnas_a_corregir:
        if col in dataframe.columns:
            
            dataframe[col] = (dataframe[col]
                              .astype(str)
                              .str.replace(diccionario_cambios, regex=True))


Llammos la funcion y con el dataframe, la lista de campos y el dicionarios con las correciones

In [511]:
lista_campos = ['Nom', 'Cognoms', 'Ciutat', 'Carrec', "País d'origen"]
corregir_subcadenas_campo(df, lista_campos, mapeo_caracteres)

Corección de Genero

Homogenizando la nomenclatura para los generos

In [512]:

df['Gènere'] = df['Gènere'].replace({'H': 'Home', 'D': 'Dona', 'F': 'Dona'})


agrupamos todos los nombres que tienen genero incorecto, haciendo dos conjuntos un para las Mujeres y otro para los Hombres

Dicionario de Generos

In [513]:
diccionario_generos = {
    # Nombres de mujer ('Dona')
    'Adri': 'Dona', 'Aina': 'Dona', 'Alba': 'Dona', 'Alexia': 'Dona', 'Anna': 'Dona', 
    'Camille': 'Dona', 'Carmen': 'Dona', 'Charlotte': 'Dona', 'Chiara': 'Dona', 
    'Clara': 'Dona', 'Claudia': 'Dona', 'Elena': 'Dona', 'Élise': 'Dona', 
    'Francesca': 'Dona', 'Gabriele': 'Dona', 'Giulia': 'Dona', 'Hannah': 'Dona', 
    'Ilaria': 'Dona', 'Irene': 'Dona', 'Joana': 'Dona', 'Katharina': 'Dona', 
    'Laia': 'Dona', 'Léa': 'Dona', 'Lucía': 'Dona', 'Marie': 'Dona', 'Marina': 'Dona', 
    'Marta': 'Dona', 'Martina': 'Dona', 'Mia': 'Dona', 'Núria': 'Dona', 'Sara': 'Dona', 
    'Simone': 'Dona', 'Sofie': 'Dona', 'Valentina': 'Dona',

    # Nombres de hombre ('Home')
    'Adrià': 'Home', 'Alejandro': 'Home', 'Alessandro': 'Home', 'Antoine': 'Home', 
    'Ben': 'Home', 'Carlos': 'Home', 'David': 'Home', 'Elias': 'Home', 'Erik': 'Home', 
    'Felix': 'Home', 'Guillem': 'Home', 'Hugo': 'Home', 'Jordi': 'Home', 'Lars': 'Home', 
    'Luca': 'Home', 'Lucas': 'Home', 'Magnus': 'Home', 'Marc': 'Home', 'Marco': 'Home', 
    'Matteo': 'Home', 'Max': 'Home', 'Nicolas': 'Home', 'Nil': 'Home', 'Noa': 'Home', 
    'Noah': 'Home', 'Oriol': 'Home', 'Paolo': 'Home', 'Pau': 'Home', 'Paul': 'Home', 
    'Pol': 'Home', 'Riccardo': 'Home', 'Sergio': 'Home', 'Thomas': 'Home', 
    'Víctor': 'Home', 'Xavier': 'Home'
}


Aplicamos el diccionario mediante map al dataframe

In [514]:
df['Gènere'] = df['Nom'].map(diccionario_generos).combine_first(df['Gènere'])

Identificacion y Eliminacion de registros duplicados

In [515]:
df_todos_duplicados = df[df.duplicated(subset=['DNI'], keep=False)]
df_duplicados_agrupados = df_todos_duplicados.sort_values(by='DNI')

print(f"Se encontraron {len(df_duplicados_agrupados)} filas asociadas a DNIs duplicados:\n")
with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'display.width', 1000):
    display(df_duplicados_agrupados)


Se encontraron 24 filas asociadas a DNIs duplicados:



,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Carrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
604,Laia,Navarro,15909601K,Espanya,Bilbao,30.0,5,1982,Dona,"1,676 €",True,False,Grup A,Analista,1.0,True,50.0,NaN,14.2
250,Laia,Navarro Vila,15909601K,Espanya,Bilbao,30.0,5,1962,Dona,1235 euros,True,True,Grup D,Analista junior,0.0,False,NaN,NaN,10.3
759,Oriol,Gómez,16618423V,Espanya,Sevilla,10.0,1,1968,Home,"2,000 €",NaN,NaN,Grup A,Responsable de vendes,5.0,NaN,NaN,NaN,20.7
758,Oriol,Gómez,16618423V,Espanya,Màlaga,10.0,1,1968,Home,1879 euros,True,NaN,Grup B,Data Analyst,4.0,False,NaN,NaN,NaN
754,Oriol,Gómez,16618423V,Espanya,Barcelona,10.0,1,1968,Home,2.358 €,True,NaN,NaN,NaN,4.0,NaN,NaN,NaN,NaN
654,Oriol,Gómez,16618423V,Espanya,Madrid,10.0,1,1968,Home,1731 euros,True,NaN,Grup B,Data analyst,4.0,NaN,NaN,NaN,NaN
692,Alexia,Pérez Díaz,49529438G,Espanya,Màlaga,12.0,10,1981,Dona,1663€,NaN,True,Grup C,Cap de projecte,NaN,True,80000.0,10.7,16.9
558,Alexia,Pérez Díaz,49529438G,Espanya,Màlaga,12.0,10,1981,Dona,"1,176 €",NaN,True,Grup A,Data Analyst,0.0,False,NaN,NaN,38.0
757,Oriol,Ferrer,52590001R,Espanya,Zaragoza,15.0,7,1973,Home,786€,NaN,True,Grup A,Analista,0.0,True,81381.0,6.5,13.9
1004,Oriol,Ferrer,52590001R,Espanya,Zaragoza,15.0,7,1973,Home,786€,NaN,True,Grup A,Analista,0.0,True,81381.0,6.5,13.9


In [326]:
df_resultado = df[df['DNI'] == '15909601K']

with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'display.width', 1000):
    print(df_resultado)

      Nom       Cognoms        DNI País d'origen  Ciutat  Dia de Naixement  Mes de Naixement  Any de Naixement Gènere Salari mensual Fills No Fills Grup Professional           Carrec  Nombre_fills Te_cotxe  Km_anuals  Consum_mitja_L_100km  Temperatura_mitjana_ciutat
250  Laia  Navarro Vila  15909601K       Espanya  Bilbao              30.0               5.0            1958.0   Dona     1235 euros  True     True            Grup D  Analista junior           0.0    False        NaN                   NaN                        10.3
604  Laia       Navarro  15909601K       Espanya  Bilbao              30.0               5.0            1958.0   Dona      1,676 €  True    False            Grup A         Analista           1.0     True       50.0                   NaN                        14.2


Con e DNI: 15909601K , 
- eliminar el registro 250 porque en es la misma persona que ha tenido hijo y ha mejorado en el cargo, ademas adquirio coche

In [516]:
df = df.drop(250)

In [189]:
df_resultado = df[df['DNI'] == '16618423V']

with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'display.width', 1000):
    print(df_resultado)

       Nom Cognoms        DNI País d'origen     Ciutat Gènere  Salari mensual Fills No Fills Grup Professional                 Carrec  Nombre_fills Te_cotxe  Km_anuals  Consum_mitja_L_100km  Temperatura_mitjana_ciutat Fecha_Nacimiento_Validada
654  Oriol   Gómez  16618423V       Espanya     Madrid   Home         1731.00  True      NaN            Grup B           Data analyst           4.0      NaN        NaN                   NaN                         NaN                1968-01-10
754  Oriol   Gómez  16618423V       Espanya  Barcelona   Home            2.36  True      NaN               NaN                    NaN           4.0      NaN        NaN                   NaN                         NaN                1968-01-10
758  Oriol   Gómez  16618423V       Espanya     Malaga   Home         1879.00  True      NaN            Grup B           Data Analyst           4.0    False        NaN                   NaN                         NaN                1968-01-10
759  Oriol   Gómez  1661

Con e DNI: 16618423V , eliminar el registro 
- Eliminamos los registros 654,754,,758, al ser la misma persona que ha cambiado de ciudad, ha tenido un hijo mas y cambiado de trabajo, por loq ue queda el registro 759 

In [517]:
df = df.drop([654,754,758])

In [191]:
df_resultado = df[df['DNI'] == '49529438G']

with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'display.width', 1000):
    print(df_resultado)

        Nom     Cognoms        DNI País d'origen  Ciutat Gènere  Salari mensual Fills No Fills Grup Professional           Carrec  Nombre_fills Te_cotxe  Km_anuals  Consum_mitja_L_100km  Temperatura_mitjana_ciutat Fecha_Nacimiento_Validada
558  Alexia  Pérez Díaz  49529438G       Espanya  Malaga   Dona          1176.0   NaN     True            Grup A     Data Analyst           0.0    False        NaN                   NaN                        38.0                1981-10-12
692  Alexia  Pérez Díaz  49529438G       Espanya  Malaga   Dona          1663.0   NaN     True            Grup C  Cap de projecte           NaN     True    80000.0                  10.7                        16.9                1981-10-12


Con e DNI: 49529438G , 
- eliminar el registro 558 porque es la mims persona que ha cambiado de cargo y ha comprado un coche

In [518]:
df = df.drop(558)

In [193]:
df_resultado = df[df['DNI'] == '52590001R']

with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'display.width', 1000):
    print(df_resultado)

        Nom Cognoms        DNI País d'origen    Ciutat Gènere  Salari mensual Fills No Fills Grup Professional    Carrec  Nombre_fills Te_cotxe  Km_anuals  Consum_mitja_L_100km  Temperatura_mitjana_ciutat Fecha_Nacimiento_Validada
757   Oriol  Ferrer  52590001R       Espanya  Zaragoza   Home           786.0   NaN     True            Grup A  Analista           0.0     True    81381.0                   6.5                        13.9                1973-07-15
1004  Oriol  Ferrer  52590001R       Espanya  Zaragoza   Home           786.0   NaN     True            Grup A  Analista           0.0     True    81381.0                   6.5                        13.9                1973-07-15


Con e DNI: 52590001R
- eliminar el registro 1004 pues los dos registros son identicos por lo que no parta nada nuevo

In [519]:
df = df.drop(1004)

In [195]:
df_resultado = df[df['DNI'] == '54013571W']

with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'display.width', 1000):
    print(df_resultado)

         Nom      Cognoms        DNI País d'origen    Ciutat Gènere  Salari mensual Fills No Fills Grup Professional        Carrec  Nombre_fills Te_cotxe  Km_anuals  Consum_mitja_L_100km  Temperatura_mitjana_ciutat Fecha_Nacimiento_Validada
89   Guillem  Pujol Costa  54013571W       Espanya  A Coruña   Home            1.57  True      NaN            Grup B  Project lead           4.0    False        NaN                   NaN                        14.6                1983-07-17
142    Marta         Ruiz  54013571W       Espanya     Palma   Dona         1276.00  True     True            Grup A      Analista           0.0    False        NaN                   NaN                        15.7                1982-05-24


Con e DNI: 54013571W
- eliminar el registro 89 porque no tiene un salario coherente.

In [520]:
df = df.drop(89)

In [197]:
df_resultado = df[df['DNI'] == '55283701C']

with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'display.width', 1000):
    print(df_resultado)

    Nom          Cognoms        DNI País d'origen Ciutat Gènere  Salari mensual Fills No Fills Grup Professional    Carrec  Nombre_fills Te_cotxe  Km_anuals  Consum_mitja_L_100km  Temperatura_mitjana_ciutat Fecha_Nacimiento_Validada
36  Ida  Karlsen Eriksen  55283701C       Noruega   Oslo   Dona          1614.0  True      NaN            Grup B  Analista           3.0      NaN        NaN                   NaN                        10.6                1992-09-02
40  Ida  Karlsen Eriksen  55283701C       Noruega   Oslo   Dona          1614.0  True      NaN            Grup B  Analista           3.0      NaN        NaN                   NaN                        10.6                1992-09-02


Con e DNI: 55283701C
- se puede eliminar cualquiera de los dos registros pue son identicos.eliminemos el registro 40

In [521]:
df = df.drop(40)

In [23]:
df_resultado = df[df['DNI'] == '55297384F']

with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'display.width', 1000):
    print(df_resultado)

          Nom     Cognoms        DNI País d'origen    Ciutat  Dia de Naixement  Mes de Naixement  Any de Naixement Gènere  Salari mensual Fills No Fills Grup Professional                 Carrec  Nombre_fills Te_cotxe  Km_anuals  Consum_mitja_L_100km  Temperatura_mitjana_ciutat Fecha_Nacimiento_Validada
730   Claudia  Gil Ferrer  55297384F       Espanya  Valencia                 1                11              1981   Dona            1.79   NaN     True            Grup B  Responsable de vendes           0.0     True    29952.0                   9.2                        20.3                1981-11-01
1003  Claudia  Gil Ferrer  55297384F       Espanya  Valencia                 1                11              1981   Dona            1.79   NaN     True            Grup B  Responsable de vendes           0.0     True    29952.0                   9.2                        20.3                1981-11-01


Con e DNI: 55297384F
- se puede eliminar cualquiera de los dos registros pues son identicos.

In [522]:
df = df.drop(1003)

In [ ]:
df_resultado = df[df['DNI'] == '90585882A']

with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'display.width', 1000):
    print(df_resultado)

Con e DNI: 90585882A
- se puede eliminar los registros 25,913 y dejar el registro 1000 poque es la misma persona y lo que se observa es un aumento de salario
-eliminaria los registro 281 y 114 porque hay valores refrenete a los hijos que estan en Nan, y el registro 104 porque si hay un evolucion en el salario de Sara Serra es porque ella esla que le pertenece ese DNI.

In [523]:
df = df.drop([25,913,281,114])

In [34]:
df_resultado = df[df['DNI'] == '97954019T']

with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'display.width', 1000):
    print(df_resultado)

       Nom Cognoms        DNI País d'origen  Ciutat  Dia de Naixement  Mes de Naixement  Any de Naixement Gènere  Salari mensual Fills No Fills Grup Professional                 Carrec  Nombre_fills Te_cotxe  Km_anuals  Consum_mitja_L_100km  Temperatura_mitjana_ciutat Fecha_Nacimiento_Validada
372   Alba   Muñoz  97954019T       Espanya  Bilbao                25                10              1983   Dona          1409.0   NaN     True            Grup B  Responsable de vendes           0.0      NaN        NaN                   NaN                        14.7                1983-10-25
1002  Alba   Muñoz  97954019T       Espanya  Bilbao                25                10              1983   Dona          1409.0   NaN     True            Grup B  Responsable de vendes           0.0      NaN        NaN                   NaN                        14.7                1983-10-25


Con e DNI: 97954019T
- se puede eliminar  cualquiera de los dos registros poque son identicos.por ejemplo el 1002

In [524]:
df = df.drop(1002)

# Ejercicio 3

 Limpieza inicial: Convertir a numérico y eliminar ceros

In [525]:
columnas_fecha = ['Any de Naixement', 'Mes de Naixement', 'Dia de Naixement']
for col in columnas_fecha:
    df[col] = pd.to_numeric(df[col], errors='coerce')
    df[col] = df[col].replace(0, np.nan)


 Corregir MESES fuera de rango (Menores a 1 o mayores a 12, y nulos)
 Reemplazamos valores fuera de rango por NaN para unificar el tratamiento

In [526]:
df.loc[(df['Mes de Naixement'] < 1) | (df['Mes de Naixement'] > 12), 'Mes de Naixement'] = np.nan

Asignamos un mes aleatorio entre 1 y 12 a los faltantes/erróneos para no acumular todo en Enero

In [527]:
meses_validos = df['Mes de Naixement'].dropna()
if meses_validos.empty:
    meses_validos = pd.Series(range(1, 13))

df['Mes de Naixement'] = df['Mes de Naixement'].apply(lambda x: int(np.random.choice(meses_validos)) if pd.isna(x) else int(x))


Corregir AÑOS (Mismo método anterior para jubilados y menores)

In [528]:
ANIO_ACTUAL = datetime.now().year
ANIO_MIN_MINIMO = ANIO_ACTUAL - 67  
ANIO_MAX_MAXIMO = ANIO_ACTUAL - 16  

anios_validos = df['Any de Naixement'].dropna()
anios_validos = anios_validos[(anios_validos >= ANIO_MIN_MINIMO) & (anios_validos <= ANIO_MAX_MAXIMO)]
if anios_validos.empty:
    anios_validos = pd.Series(range(ANIO_MIN_MINIMO, ANIO_MAX_MAXIMO + 1))

df['Any de Naixement'] = df['Any de Naixement'].apply(
    lambda x: int(np.random.choice(anios_validos)) if (pd.isna(x) or x < ANIO_MIN_MINIMO or x > ANIO_MAX_MAXIMO) else int(x)
)


Corregir DÍAS fuera de rango según el mes y año (Manejo de bisiestos)
Diccionario básico de días máximos por mes (Febrero se calcula dinámicamente después)
dias_por_mes = {1:31, 2:28, 3:31, 4:30, 5:31, 6:30, 7:31, 8:31, 9:30, 10:31, 11:30, 12:31}

Calcular el máximo de días para el mes actual
Si es bisiesto, Febrero tiene 29 días
Si el día supera el máximo permitido para ese mes, lo forzamos a 1

In [529]:
def validar_dia(row):
    dia = row['Dia de Naixement']
    mes = row['Mes de Naixement']
    anio = row['Any de Naixement']
    
    if pd.isna(dia) or dia < 1:
        return 1
    
   
    max_dia = dias_por_mes.get(mes, 31)
    if mes == 2:
        
        if (anio % 4 == 0 and anio % 100 != 0) or (anio % 400 == 0):
            max_dia = 29
            
    
    if dia > max_dia:
        return 1
    return int(dia)

df['Dia de Naixement'] = df.apply(validar_dia, axis=1).astype(int)


Crear la fecha final validada 

In [530]:
df['Fecha_Nacimiento_Validada'] = pd.to_datetime(dict(
    year=df['Any de Naixement'], 
    month=df['Mes de Naixement'], 
    day=df['Dia de Naixement']
), errors='coerce')


eliminar colunnas  originals con datos de fechas

In [531]:
columnas_a_eliminar = ["Any de Naixement", "Mes de Naixement", "Dia de Naixement"]
df = df.drop(columns=columnas_a_eliminar, errors="ignore")

Analisis de las columnas relacionadas con los hijos

 Limpiar la columna numérica 'nombre de filis' rellenando con 0
  Si 'No Fills' es True o 'Fills' es False, asumimos que tiene 0 hijos de forma segura

In [532]:
condicion_cero_hijos = (df["No Fills"] == True) | (df["Fills"] == False)
df.loc[condicion_cero_hijos, "Nombre_fills"] = df.loc[condicion_cero_hijos, "Nombre_fills"].fillna(0)

Crear la columna final unificada 'Fills'
Por defecto, tomamos el valor numérico que ya existe

In [533]:
df["Fills_ok"] = df["Nombre_fills"]

 Rectificar los NaN restantes cruzando las columnas booleanas
Si 'Fills' es True pero el número de hijos es NaN, asumimos al menos 1 hijo 

In [534]:
df.loc[(df["Fills"] == True) & (df["Nombre_fills"].isna()), "Fills"] = 1

Si todo sigue siendo NaN (no hay info en ninguna columna), asumimos 0 por defecto para limpiar el dataset

In [535]:
df["Fills_ok"] = df["Fills_ok"].fillna(0)

Convertir a entero (int) para eliminar los decimales (.0) 

In [536]:
df["Fills_ok"] = df["Fills_ok"].astype(int)


Eliminar las 3 columnas que ya no necesitamos para dejar el dataset limpio

In [537]:
columnas_a_eliminar = ["Fills", "No Fills", "Nombre_fills"]
df = df.drop(columns=columnas_a_eliminar, errors="ignore")

Analisis de datos coches

Si hay valores negativos por error al transcribir lo convertimos a positivo positivo

In [538]:
df["Km_anuals"] = df["Km_anuals"].abs()

Si 'Km_anuals' o 'Consum_mitja_L_100km' no son vacios entonces gaarantizamos que tiene coche

In [539]:
tiene_datos_coche = df['Km_anuals'].notna() | df['Consum_mitja_L_100km'].notna()
df.loc[tiene_datos_coche, 'Te_cotxe'] = True


Resolución de incoherencia (Consumo sin Kilometraje): Imputamos el Km mediante la mediana 
de los que sí tienen coche para no alterar la distribución con valores extremos (outliers)

In [552]:
incoherencia_km = df['Consum_mitja_L_100km'].notna() & df['Km_anuals'].isna()
mediana_km = df[df['Te_cotxe'] == True]['Km_anuals'].median()
df.loc[incoherencia_km, 'Km_anuals'] = round(mediana_km,1)


 Resolución de incoherencia (Kilometraje sin consumo): Imputamos el consumo mediante la mediana 
 de los que sí tienen coche para no alterar la distribución con valores extremos (outliers)

In [541]:
incoherencia_consum_L = df['Consum_mitja_L_100km'].isna() & df['Km_anuals'].notna()
mediana_consum_L = df[df['Te_cotxe'] == True]['Consum_mitja_L_100km'].median()
df.loc[incoherencia_consum_L, 'Consum_mitja_L_100km'] = round(mediana_consum_L,1)

Si no tiene datos de coche en absoluto, aseguramos que conste como 'False' y sus métricas sean 0

In [543]:
df.loc[df['Te_cotxe'].isna(), 'Te_cotxe'] = False
df.loc[df['Te_cotxe'] == False, ['Km_anuals', 'Consum_mitja_L_100km']] = 0

Identificamos a las personas que tienen 'Te_cotxe' == TRUE pero sus métricas están vacías

In [544]:
incoherencia_tiene_coche = (df['Te_cotxe'] == True) & (df['Km_anuals'].isna()) & (df['Consum_mitja_L_100km'].isna())

Calculamos las medianas de referencia del grupo que SÍ tiene coche y SÍ tiene datos válidos

In [545]:
mediana_km_con_coche = df[(df['Te_cotxe'] == True) & (df['Km_anuals'] > 0)]['Km_anuals'].median()
mediana_consumo_con_coche = df[(df['Te_cotxe'] == True) & (df['Consum_mitja_L_100km'] > 0)]['Consum_mitja_L_100km'].median()

Asignamos estas medianas solo a los registros que presentaban la incoherencia

In [546]:
df.loc[incoherencia_tiene_coche, 'Km_anuals'] = mediana_km_con_coche
df.loc[incoherencia_tiene_coche, 'Consum_mitja_L_100km'] = mediana_consumo_con_coche

La Temperatura

 IMPUTACIÓN DE TEMPERATURA MEDIA

Agrupamos por 'Ciutat' para obtener su temperatura media. Si la ciudad no tiene datos, 
usamos una jerarquía superior: la media global del dataset.

In [547]:

if 'Temperatura_mitjana_ciutat' in df.columns:
    media_por_ciudad = df.groupby('Ciutat')['Temperatura_mitjana_ciutat'].transform('mean').round(1)
    media_global_temp = round(df['Temperatura_mitjana_ciutat'].mean(),1)
    
    df['Temperatura_mitjana_ciutat'] = df['Temperatura_mitjana_ciutat'].fillna(media_por_ciudad).fillna(media_global_temp)

IMPUTACIÓN DE SALARIOS FALTANTES

Eliminar todo lo que no sea númerico 

In [548]:
df['Salari mensual'] = df['Salari mensual'].astype(str).str.replace(r'[^0-9.]', '', regex=True)
df['Salari mensual'] = pd.to_numeric(df['Salari mensual'], errors='coerce').round(2)

El salario depende estrechamente del Cargo ('Carrec') y de la 'Ciutat'. 
Calculamos la mediana de estas agrupaciones para evitar que los sueldos extremadamente altos distorsionen la imputación.

 LIMPIEZA: Convertir salarios sospechosamente bajos (ej. < 600) en NaN, para que no afectaen mis calculos de las medianas
 Calcular las medianas limpias 
 Imputación en cascada

In [549]:

if 'Salari mensual' in df.columns:
        
    UMBRAL_MINIMO = 600
    df.loc[df['Salari mensual'] < UMBRAL_MINIMO, 'Salari mensual'] = None
    
    
    mediana_compleja = df.groupby(['Carrec', 'Ciutat'])['Salari mensual'].transform('median')
    mediana_por_cargo = df.groupby('Carrec')['Salari mensual'].transform('median')
    mediana_global = df['Salari mensual'].median()
    

    df['Salari mensual'] = (df['Salari mensual'].fillna(mediana_compleja).fillna(mediana_por_cargo).fillna(mediana_global))

In [ ]:
display (df)

# Ejercio 4

Tabla resumen descriptiva

In [ ]:

tabla_genero = df.groupby('Gènere')['Salari mensual'].agg(
    Sueldo_Medio='mean',
    Mediana='median',
    Mínimo='min',
    Máximo='max'
).sort_values(by='Sueldo_Medio', ascending=False)

print("--- Resumen de Salarios por Género ---")
print(tabla_genero.round(2))


--- Resumen de Salarios por Género ---
        Sueldo_Medio  Mediana  Mínimo  Máximo
Gènere                                       
Home         1532.97   1377.0   673.0  3672.0
Dona         1530.64   1393.0   722.0  3672.0


Tabla pivote con los márgenes (totales/promedios globales)

In [558]:
matriz_salario = df.pivot_table(
    values='Salari mensual', 
    index='Gènere', 
    columns="País d'origen", 
    aggfunc='mean', 
    margins=True, 
    margins_name='Media Global'
)


Aplicar formato condicional con un mapa de color (gradiente)
Esto se visualizará con colores intensos en los valores altos

In [ ]:
matriz_estilizada = matriz_salario.style.background_gradient(
    cmap='YlOrRd', 
    subset=(matriz_salario.index[:-1], matriz_salario.columns[:-1]) 
).format("{:.2f}")


matriz_estilizada

País d'origen,Alemanya,Espanya,Francia,Itàlia,Noruega,Media Global
Gènere,,,,,,
Dona,1462.63,1540.02,1391.76,1499.55,1843.85,1530.64
Home,1643.33,1529.49,1532.44,1594.73,1385.22,1532.97
Media Global,1530.40,1534.28,1470.28,1539.21,1598.53,1531.86
